pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -0,0 +1,511 @@
+		tmpData:   newTmpData(),
+	}
+
+	return b
+}
+
+func getPodName(pod *v1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func getPVCName(pvc *v1.PersistentVolumeClaim) string {
+	return pvc.Namespace + "/" + pvc.Name
+}
+
+func (b *volumeBinder) isVolumeBound(namespace string, vol *v1.Volume, checkFullyBound bool) (bool, *v1.PersistentVolumeClaim, error) {
+	if vol.PersistentVolumeClaim == nil {
+		return true, nil, nil
+	}
+
+	pvcName := vol.PersistentVolumeClaim.ClaimName
+	pvc, err := b.pvcCache.PersistentVolumeClaims(namespace).Get(pvcName)
+	if err != nil || pvc == nil {
+		return false, nil, fmt.Errorf("error getting PVC %q: %v", pvcName, err)
+	}
+
+	pvName := pvc.Spec.VolumeName
+	if pvName != "" {
+		if checkFullyBound {
+			if metav1.HasAnnotation(pvc.ObjectMeta, annBindCompleted) {
+				glog.V(5).Infof("PVC %q is fully bound to PV %q", getPVCName(pvc), pvName)
+				return true, pvc, nil
+			} else {
+				glog.V(5).Infof("PVC %q is not fully bound to PV %q", getPVCName(pvc), pvName)
+				return false, pvc, nil
+			}
+		}
+		glog.V(5).Infof("PVC %q is bound or prebound to PV %q", getPVCName(pvc), pvName)
+		return true, pvc, nil
+	}
+
+	glog.V(5).Infof("PVC %v/%v is not bound", namespace, pvc.Name)
+	return false, pvc, nil
+}
+
+// arePodVolumesBound returns true if all volumes are fully bound
+func (b *volumeBinder) arePodVolumesBound(pod *v1.Pod) bool {
+	for _, vol := range pod.Spec.Volumes {
+		if isBound, _, _ := b.isVolumeBound(pod.Namespace, &vol, true); !isBound {
+			// Pod has at least one PVC that needs binding
+			return false
+		}
+	}
+	return true
+}
+
+// getPodVolumes returns a pod's PVCs separated into bound (including prebound), unbound with delayed binding,
+// and unbound with immediate binding
+func (b *volumeBinder) getPodVolumes(pod *v1.Pod) (boundClaims []*v1.PersistentVolumeClaim, unboundClaims []*bindingInfo, unboundClaimsImmediate []*v1.PersistentVolumeClaim, err error) {
+
+	boundClaims = []*v1.PersistentVolumeClaim{}
+	unboundClaimsImmediate = []*v1.PersistentVolumeClaim{}
+	unboundClaims = []*bindingInfo{}
+
+	for _, vol := range pod.Spec.Volumes {
+		volumeBound, pvc, err := b.isVolumeBound(pod.Namespace, &vol, false)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if pvc == nil {
+			continue
+		}
+		if volumeBound {
+			boundClaims = append(boundClaims, pvc)
+		} else {
+			delayBinding := b.ctrl.shouldDelayBinding(pvc)
+			if delayBinding {
+				// Scheduler path
+				unboundClaims = append(unboundClaims, &bindingInfo{pvc: pvc})
+			} else {
+				// Immediate binding should have already been bound
+				unboundClaimsImmediate = append(unboundClaimsImmediate, pvc)
+			}
+		}
+	}
+	return boundClaims, unboundClaims, unboundClaimsImmediate, nil
+}
+
+// FindPodVolumes internally caches the matching PVs per node in tmpData.nodeBindings
+func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, nodeName string) (needsBinding bool, foundMatches bool, err error) {
```

Comment:
Try splitting this up into 3 separate calls to not overload return values.

Reply:
Or what about these as return values:
* boundVolumesSatisfied bool
* unboundVolumesSatisfied bool 
* err error